### PR TITLE
wifi: hostap: Avoid using internal supplicant APIs from other modules

### PR DESCRIPTION
**Pull request by rado17**

### CODE REVIEW

**Comment by maxd-nordic** [comment]
thanks for working on this! it will be very useful.

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c [hunk 1]
```diff
@@ -24,6 +36,69 @@ LOG_MODULE_REGISTER(wifi_mgmt_ext, CONFIG_WIFI_MGMT_EXT_LOG_LEVEL);
...
+	params->ssid = ssid;
+	params->ssid_length = creds->header.ssid_len;
+
+	/* PSK (optional) */
+	if (creds->password_len > 0) {
+		psk = (char *)k_malloc(creds->password_len + 1);
+		if (!psk) {
+			LOG_ERR("Failed to allocate memory for PSK\n");
+			ret = -ENOMEM;
+		goto no_mem_out;
```
Comment (sachinthegreen) [inline subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c:64]: Indentation

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c [hunk 2]
```diff
@@ -24,6 +36,69 @@ LOG_MODULE_REGISTER(wifi_mgmt_ext, CONFIG_WIFI_MGMT_EXT_LOG_LEVEL);
...
 
+static int __stored_creds_to_params(struct wifi_credentials_personal *creds,
+				struct wifi_connect_req_params *params)
+{
+	char *ssid = NULL;
+	char *psk = NULL;
+	int ret;
+
+	/* SSID */
+	ssid = (char *) k_malloc(creds->header.ssid_len + 1);
```
Comment (sachinthegreen) [inline subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c:47]: No space between `(char *)`  and `k_malloc`

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c [hunk 3]
```diff
@@ -156,25 +213,59 @@ static int add_static_network_config(void)
...
-	return add_network_from_credentials_struct_personal(&creds);
+	return add_network_from_credentials_struct_personal(&creds, iface);
 #else
 	return 0;
 #endif /* defined(CONFIG_WIFI_CREDENTIALS_STATIC) */
 }
 
+static void wifi_mgmt_ext_connect_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status = (const struct wifi_status *) cb->info;
```
Comment (sachinthegreen) [inline subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c:224]: No space between `(const struct wifi_status *)` and  `cb->info`

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c (hunk 1 above)
Comment (rado17) [reply subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c]: Done

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c (hunk 2 above)
Comment (rado17) [reply subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c]: Done

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c (hunk 3 above)
Comment (rado17) [reply subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c]: Done

File: modules/hostap/src/supp_api.c [hunk 4]
```diff
@@ -239,6 +239,18 @@ int z_wpa_supplicant_connect(const struct device *dev,
 	_wpa_cli_cmd_v("set_network %d scan_ssid 1", resp.network_id);
 	_wpa_cli_cmd_v("set_network %d key_mgmt NONE", resp.network_id);
 	_wpa_cli_cmd_v("set_network %d ieee80211w 0", resp.network_id);
+	if ((params->band & BIT(WIFI_FREQ_BAND_2_4_GHZ)) == BIT(WIFI_FREQ_BAND_2_4_GHZ)) {
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:242]: This capability of setting channels, should be in a separate commit.

Comment (krish2718) [reply modules/hostap/src/supp_api.c:242]: And have you tried skipping this altogether and let WPA supplicant derive these from the driver's capability? Its not ideal to duplicate these esp. without the regulatory domain information.

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c [hunk 5]
```diff
@@ -24,6 +24,60 @@ LOG_MODULE_REGISTER(wifi_mgmt_ext, CONFIG_WIFI_MGMT_EXT_LOG_LEVEL);
...
+		return -ENOMEM;
+	}
+	memset(ssid, 0, creds->header.ssid_len + 1);
+	strncpy(ssid, creds->header.ssid, creds->header.ssid_len);
+	params->ssid = ssid;
+	params->ssid_length = creds->header.ssid_len;
+
+	/* PSK (optional) */
+	if (creds->password_len > 0) {
+		psk = (char *)k_malloc(creds->password_len + 1);
```
Comment (krish2718) [inline subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c:60]: use `k_zalloc` to avoid extra memset.

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c [hunk 6]
```diff
@@ -24,6 +24,60 @@ LOG_MODULE_REGISTER(wifi_mgmt_ext, CONFIG_WIFI_MGMT_EXT_LOG_LEVEL);
...
+		if (!psk) {
+			LOG_ERR("Failed to allocate memory for PSK\n");
+			return -ENOMEM;
+		}
+		memset(psk, 0, creds->password_len + 1);
+		strncpy(psk, creds->password, creds->password_len);
```
Comment (krish2718) [inline subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c:67]: no check for truncation and missing NULL termination?

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c [hunk 7]
```diff
@@ -24,6 +24,60 @@ LOG_MODULE_REGISTER(wifi_mgmt_ext, CONFIG_WIFI_MGMT_EXT_LOG_LEVEL);
...
+	/* Defaults */
+	params->security = WIFI_SECURITY_TYPE_PSK;
+	params->mfp = WIFI_MFP_OPTIONAL;
+	params->channel = WIFI_CHANNEL_ANY;
+	params->timeout = CONNECTION_TIMEOUT;
+
+	/* Security type (optional) */
+	if (creds->header.type <= WIFI_SECURITY_TYPE_MAX) {
+		params->security = creds->header.type;
+	} else {
```
Comment (krish2718) [inline subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c:81]: The current default is overwritten in both branches, so, if you make this as default then `else` branch can be removed.

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: modules/hostap/src/supp_api.c (hunk 4 above)
Comment (rado17) [reply modules/hostap/src/supp_api.c:242]: Should we then remove band configuration support from stored wifi credentials as well?

Comment (krish2718) [reply modules/hostap/src/supp_api.c:242]: Well, we can still keep it and implement, but we need to dynamically query the supported channels from the driver and use that rather than a static list.

**Review by sachinthegreen — APPROVED** [review]

**Review by kapbh — APPROVED** [review]

File: modules/hostap/src/supp_api.c (hunk 4 above)
Comment (rado17) [reply modules/hostap/src/supp_api.c:250]: Done

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c (hunk 5 above)
Comment (rado17) [reply subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c:66]: Looks like k_zalloc is not available in zephyr. Keeping this as is.

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c (hunk 6 above)
Comment (rado17) [reply subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c]: Moved to snprintf which takes care of adding the null termination and added checks for truncation.

File: subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c (hunk 7 above)
Comment (rado17) [reply subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c]: Done

File: modules/hostap/src/supp_api.c [hunk 8]
```diff
@@ -242,6 +247,67 @@ int z_wpa_supplicant_connect(const struct device *dev,
...
+				retval = snprintf(chan_list + offset, 6, " %d",
+						     mode->channels[i].freq);
+				offset += retval;
+			}
+
+			printf("chanlist_24: %s\n", chan_list);
+			_wpa_cli_cmd_v("set_network %d freq_list%s", resp.network_id, chan_list);
+			k_free(chan_list);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (rado17) [reply modules/hostap/src/supp_api.c]: Done.

File: modules/hostap/src/supp_api.c [hunk 9]
```diff
@@ -144,6 +144,73 @@ static void supp_shell_connect_status(struct k_work *work)
...
+
+			retval = 0;
+			offset = 0;
+			for (i = 0; i < mode->num_channels; i++) {
+				retval = snprintf(_chan_list + offset, 6, " %d",
+						  mode->channels[i].freq);
+				offset += retval;
+			}
+			*chan_list = _chan_list;
+		}
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:180]: This logic is common for 2.4/5, so, can be refactored?

Comment (rado17) [reply modules/hostap/src/supp_api.c]: Done

File: modules/hostap/src/supp_api.c [hunk 10]
```diff
@@ -144,6 +144,52 @@ static void supp_shell_connect_status(struct k_work *work)
...
+	struct hostapd_hw_modes *mode = NULL;
+	int i, j;
+	int offset, retval;
+	int size;
+	char *_chan_list;
+
+	if ((band & BIT(WIFI_FREQ_BAND_2_4_GHZ)) == BIT(WIFI_FREQ_BAND_2_4_GHZ)) {
+		mode = NULL;
+		for (j = 0; j < ((wpa_s)->hw).num_modes; j++) {
+			if (((wpa_s)->hw).modes[j].mode != HOSTAPD_MODE_IEEE80211A) {
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:158]: This assumes dual band only, what about 6GHz?

File: modules/hostap/src/supp_api.c [hunk 11]
```diff
@@ -144,6 +144,52 @@ static void supp_shell_connect_status(struct k_work *work)
...
+			}
+		}
+	} else if ((band & BIT(WIFI_FREQ_BAND_5_GHZ)) == BIT(WIFI_FREQ_BAND_5_GHZ)) {
+		mode = NULL;
+		for (j = 0; j < ((wpa_s)->hw).num_modes; j++) {
+			if (((wpa_s)->hw).modes[j].mode == HOSTAPD_MODE_IEEE80211A) {
+				mode = &((wpa_s)->hw).modes[j];
+				break;
+			}
+		}
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:170]: Even this can be refactored if you can get mode i.e., `HOSTAPD_MODE_*` from `WIFI_FREQ_BAND_*`.

File: modules/hostap/src/supp_api.c [hunk 12]
```diff
@@ -144,6 +144,52 @@ static void supp_shell_connect_status(struct k_work *work)
...
+	}
+
+	if (mode) {
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:173]: Return early to save indent.

```suggestion
	if (!mode) {
	    return 0;
	}
```

File: subsys/net/lib/wifi_credentials/Kconfig [hunk 13]
```diff
@@ -40,7 +40,7 @@ endchoice
 
 config WIFI_CREDENTIALS_MAX_ENTRIES
 	int "Number of supported WiFi credentials"
-	default 2
+	default 1
```
Comment (krish2718) [inline subsys/net/lib/wifi_credentials/Kconfig:43]: > Hence reducing the max entries to 1 to avoid these 2 issues.

This sounds like a workaround than actual fix, what if the user configures this to 2?

**Comment by krish2718** [comment]
> attempt for connection is initiated with the subsequent credentials in the stored list.

This doesn't seem right, second credentials should only be used if first one isn't successful.

> Under DISCONNECTED event processing, l2_conn module
tries to read the flags which again is done under conn_binding lock
which causes a deadlock.

This sounds like a bug in the code, we should fix the locking.

File: modules/hostap/src/supp_api.c (hunk 10 above)
Comment (rado17) [reply modules/hostap/src/supp_api.c]: Added checks for 6GHz

File: modules/hostap/src/supp_api.c (hunk 12 above)
Comment (rado17) [reply modules/hostap/src/supp_api.c]: Done

File: modules/hostap/src/supp_api.c (hunk 11 above)
Comment (rado17) [reply modules/hostap/src/supp_api.c]: Made changes to identify hardware mode based on band.

File: modules/hostap/src/supp_api.c [hunk 14]
```diff
@@ -145,6 +145,53 @@ static void supp_shell_connect_status(struct k_work *work)
...
+	int i;
+	int offset, retval;
+	int size;
+	char *_chan_list;
+
+	mode = get_mode_by_band(wpa_s, band);
+	if (!mode) {
+		return -EINVAL;
+	}
+	size = ((mode->num_channels) * 6) + 1;
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:177]: 6?

File: modules/hostap/src/supp_api.c [hunk 15]
```diff
@@ -145,6 +145,53 @@ static void supp_shell_connect_status(struct k_work *work)
 	k_mutex_unlock(&wpa_supplicant_mutex);
 }
 
+static struct hostapd_hw_modes *get_mode_by_band(struct wpa_supplicant *wpa_s, uint8_t band)
+{
+	enum hostapd_hw_mode hw_mode;
+	bool is_6ghz = (band & BIT(WIFI_FREQ_BAND_6_GHZ))?true:false;
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:151]: Ternary needs spaces around

File: modules/hostap/src/supp_api.c [hunk 16]
```diff
@@ -145,6 +145,53 @@ static void supp_shell_connect_status(struct k_work *work)
...
+	int offset, retval;
+	int size;
+	char *_chan_list;
+
+	mode = get_mode_by_band(wpa_s, band);
+	if (!mode) {
+		return -EINVAL;
+	}
+	size = ((mode->num_channels) * 6) + 1;
+	_chan_list = (char *)k_malloc(size);
```
[the remark this answers is not in the document]
Comment (rado17) [reply modules/hostap/src/supp_api.c]: Removed the casting

File: modules/hostap/src/supp_api.c (hunk 14 above)
Comment (rado17) [reply modules/hostap/src/supp_api.c]: Done

File: modules/hostap/src/supp_api.c [hunk 17]
```diff
@@ -145,6 +145,53 @@ static void supp_shell_connect_status(struct k_work *work)
 	k_mutex_unlock(&wpa_supplicant_mutex);
 }
 
+static struct hostapd_hw_modes *get_mode_by_band(struct wpa_supplicant *wpa_s, uint8_t band)
+{
+	enum hostapd_hw_mode hw_mode;
+	bool is_6ghz = (band & BIT(WIFI_FREQ_BAND_6_GHZ)) ? true:false;
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:151]: `(band & BIT(WIFI_FREQ_BAND_6_GHZ)) ? true : false;`

File: modules/hostap/src/supp_api.c [hunk 18]
```diff
@@ -145,6 +145,53 @@ static void supp_shell_connect_status(struct k_work *work)
...
+
+	if ((band & BIT(WIFI_FREQ_BAND_2_4_GHZ))) {
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:153]: Just realized that the Wi-Fi management bands are not bitmaps, just normal enum values.

File: modules/hostap/src/supp_api.c [hunk 19]
```diff
@@ -145,6 +145,53 @@ static void supp_shell_connect_status(struct k_work *work)
...
+	int i;
+	int offset, retval;
+	int size;
+	char *_chan_list;
+
+	mode = get_mode_by_band(wpa_s, band);
+	if (!mode) {
+		return -EINVAL;
+	}
+	size = ((mode->num_channels) * CHAN_NUM_LEN) + 1;
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:177]: a newline before

File: modules/hostap/src/supp_api.h [hunk 20]
```diff
@@ -12,6 +12,7 @@
 
 #define MAX_SSID_LEN 32
 #define MAC_ADDR_LEN 6
+#define CHAN_NUM_LEN 6
```
Comment (krish2718) [inline modules/hostap/src/supp_api.h:15]: `x.abc` + NULL terminator?

Comment (rado17) [reply modules/hostap/src/supp_api.h:15]: abcd+" "+NULL terminator

File: modules/hostap/src/supp_api.c (hunk 18 above)
Comment (rado17) [reply modules/hostap/src/supp_api.c:153]: Yes. Used BIT macro here to differentiate the case where no band is assigned (band initialized to 0 in configuration API) and the case where band is set to 2.4GHz.

File: modules/hostap/src/supp_api.c (hunk 19 above)
Comment (rado17) [reply modules/hostap/src/supp_api.c:178]: Done

File: modules/hostap/src/supp_api.c (hunk 17 above)
Comment (rado17) [reply modules/hostap/src/supp_api.c]: Done.

File: modules/hostap/src/supp_api.c (hunk 18 above)
Comment (krish2718) [reply modules/hostap/src/supp_api.c:153]: But the input is not modified `params->band` is a normal value not BIT()?

File: modules/hostap/src/supp_api.h (hunk 20 above)
Comment (krish2718) [reply modules/hostap/src/supp_api.h:15]: May be adding a comment might help including `space is needed to concatenate`

File: modules/hostap/src/supp_api.c (hunk 18 above)
Comment (rado17) [reply modules/hostap/src/supp_api.c:153]: We are setting BIT() for param band in wifi_ext_mgmt module:

```
if (creds->header.flags & WIFI_CREDENTIALS_FLAG_2_4GHz) {
		params->band |= BIT(WIFI_FREQ_BAND_2_4_GHZ);
	}

	if (creds->header.flags & WIFI_CREDENTIALS_FLAG_5GHz) {
		params->band |= BIT(WIFI_FREQ_BAND_5_GHZ);
	}
```

File: modules/hostap/src/supp_api.h (hunk 20 above)
Comment (rado17) [reply modules/hostap/src/supp_api.h:15]: Will do

File: modules/hostap/src/supp_api.c (hunk 18 above)
Comment (krish2718) [reply modules/hostap/src/supp_api.c:153]: That is wrong, as per docs `params->band` is just an enum, so, any driver will try to process it as enum and might fail. Though I don't see this param used currently.

Comment (rado17) [reply modules/hostap/src/supp_api.c:153]: Do you see any issue if I change the enum definition i.e., set 2,4GHz to 1?

Comment (krish2718) [reply modules/hostap/src/supp_api.c:153]: Nope, its fine. But that would mean an upstream change :)

Comment (rado17) [reply modules/hostap/src/supp_api.c:153]: I would still suggest that we use the BIT() approach, since having 0 as a valid value for a field in structure seems problematic. Specially in this case, since, for a structure instantiated to 0, the default band becomes 2.4GHz-only. 
Also, this avoids the upstream change.

Comment (krish2718) [reply modules/hostap/src/supp_api.c:153]: Sure, in that case then please use an intermediate variable rather than modify `params->band`, in fact `params` should be treated as `const` (missed this in the original implementation).

Comment (rado17) [reply modules/hostap/src/supp_api.c]: As discussed, avoided usage of `BIT()` for `params->band`. Instead setting band to `WIFI_FREQ_BAND_UNKNOWN` in case band is not configure.

File: modules/hostap/src/supp_api.c [hunk 21]
```diff
@@ -145,6 +145,54 @@ static void supp_shell_connect_status(struct k_work *work)
 	k_mutex_unlock(&wpa_supplicant_mutex);
 }
 
+static struct hostapd_hw_modes *get_mode_by_band(struct wpa_supplicant *wpa_s, uint8_t band)
+{
+	enum hostapd_hw_mode hw_mode;
+	bool is_6ghz = (band & BIT(WIFI_FREQ_BAND_6_GHZ)) ? true : false;
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:151]: This still has BIT() .

File: modules/hostap/src/supp_api.h (hunk 20 above)
Comment (krish2718) [reply modules/hostap/src/supp_api.h:15]: Don't see the comment?

Comment (rado17) [reply modules/hostap/src/supp_api.h]: Added now.

File: modules/hostap/src/supp_api.c (hunk 21 above)
Comment (rado17) [reply modules/hostap/src/supp_api.c]: Corrected this.
